ZST DST PME when calling `FromBytes::ref_from_bytes` in client code

Hello,

I am migrating from v0.6.6 to v0.8.3 but I am facing a compilation error in zerocode. The library code happily compiles on it's own but when I try to build a fuzz target it fails to compile zerocopy.

This makes me think that I have made a configuration or environmental mistake for the fuzz harness. 

However I am having issues tracing the root cause, as the error occurs when compiling the zerocopy crate. It looks like a static assert is being hit, but I'm not entirely sure what it's purpose is and how to trace it back to my code change.

Can you help me identify how to trouble shoot further?

Here is the error I am seeing:
```
error[E0080]: evaluation of `<[()] as FromBytes::ref_from_bytes::StaticAssert>::ASSERT` failed
    --> /usr/local/google/home/clundin/.cargo/registry/src/index.crates.io-6f17d22bba15001f/zerocopy-0.8.3/src/util/macros.rs:664:9
     |
664  |         assert!($e);
     |         ^^^^^^^^^^^ the evaluated program panicked at 'assertion failed: {
    let dst_is_zst =
        match Self::LAYOUT.size_info {
            crate::SizeInfo::Sized { .. } => false,
            crate::SizeInfo::SliceDst(TrailingSliceLayout { elem_size, .. })
                => {
                elem_size == 0
            }
        };
    !dst_is_zst
}', /usr/local/google/home/clundin/.cargo/registry/src/index.crates.io-6f17d22bba15001f/zerocopy-0.8.3/src/lib.rs:3508:9
     |
    ::: /usr/local/google/home/clundin/.cargo/registry/src/index.crates.io-6f17d22bba15001f/zerocopy-0.8.3/src/lib.rs:3508:9
     |
3508 |         static_assert_dst_is_not_zst!(Self);
     |         ----------------------------------- in this macro invocation
     |
     = note: this error originates in the macro `assert` which comes from the expansion of the macro `static_assert_dst_is_not_zst` (in Nightly builds, run with -Z macro-backtrace for more info)

note: erroneous constant encountered
    --> /usr/local/google/home/clundin/.cargo/registry/src/index.crates.io-6f17d22bba15001f/zerocopy-0.8.3/src/util/macros.rs:724:23
     |
724  |         const_assert!(<Self as StaticAssert>::ASSERT);
     |                       ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
     |
    ::: /usr/local/google/home/clundin/.cargo/registry/src/index.crates.io-6f17d22bba15001f/zerocopy-0.8.3/src/lib.rs:3508:9
     |
3508 |         static_assert_dst_is_not_zst!(Self);
     |         ----------------------------------- in this macro invocation
     |
     = note: this note originates in the macro `static_assert` which comes from the expansion of the macro `static_assert_dst_is_not_zst` (in Nightly builds, run with -Z macro-backtrace for more info)

note: the above error was encountered while instantiating `fn <[()] as FromBytes>::ref_from_bytes`
    --> /usr/local/google/home/clundin/.cargo/registry/src/index.crates.io-6f17d22bba15001f/zerocopy-0.8.3/src/lib.rs:4555:9
     |
4555 |         <[Self]>::ref_from_bytes(source).ok()
     |         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

error[E0080]: evaluation of `<[()] as FromBytes::mut_from_bytes::StaticAssert>::ASSERT` failed
    --> /usr/local/google/home/clundin/.cargo/registry/src/index.crates.io-6f17d22bba15001f/zerocopy-0.8.3/src/util/macros.rs:664:9
     |
664  |         assert!($e);
     |         ^^^^^^^^^^^ the evaluated program panicked at 'assertion failed: {
    let dst_is_zst =
        match Self::LAYOUT.size_info {
            crate::SizeInfo::Sized { .. } => false,
            crate::SizeInfo::SliceDst(TrailingSliceLayout { elem_size, .. })
                => {
                elem_size == 0
            }
        };
    !dst_is_zst
}', /usr/local/google/home/clundin/.cargo/registry/src/index.crates.io-6f17d22bba15001f/zerocopy-0.8.3/src/lib.rs:3744:9
     |
    ::: /usr/local/google/home/clundin/.cargo/registry/src/index.crates.io-6f17d22bba15001f/zerocopy-0.8.3/src/lib.rs:3744:9
     |
3744 |         static_assert_dst_is_not_zst!(Self);
     |         ----------------------------------- in this macro invocation
     |
     = note: this error originates in the macro `assert` which comes from the expansion of the macro `static_assert_dst_is_not_zst` (in Nightly builds, run with -Z macro-backtrace for more info)

note: erroneous constant encountered
    --> /usr/local/google/home/clundin/.cargo/registry/src/index.crates.io-6f17d22bba15001f/zerocopy-0.8.3/src/util/macros.rs:724:23
     |
724  |         const_assert!(<Self as StaticAssert>::ASSERT);
     |                       ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
     |
    ::: /usr/local/google/home/clundin/.cargo/registry/src/index.crates.io-6f17d22bba15001f/zerocopy-0.8.3/src/lib.rs:3744:9
     |
3744 |         static_assert_dst_is_not_zst!(Self);
     |         ----------------------------------- in this macro invocation
     |
     = note: this note originates in the macro `static_assert` which comes from the expansion of the macro `static_assert_dst_is_not_zst` (in Nightly builds, run with -Z macro-backtrace for more info)

note: the above error was encountered while instantiating `fn <[()] as FromBytes>::mut_from_bytes`
    --> /usr/local/google/home/clundin/.cargo/registry/src/index.crates.io-6f17d22bba15001f/zerocopy-0.8.3/src/lib.rs:4588:9
     |
4588 |         <[Self]>::mut_from_bytes(source).ok()
     |         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
````

For context, here is my change: https://github.com/chipsalliance/caliptra-dpe/pull/360

And finally here are some steps to repro locally:

```
$ DATE="2023-11-16"
$ git clone https://github.com/clundin25/caliptra-dpe --branch zero-copy-update 
$ cd caliptra-dpe/dpe/fuzz 
$ rustup install "nightly-${DATE}" 
$ cargo +"nightly-$DATE" install cargo-fuzz cargo-afl 
$ cargo +"nightly-$DATE" fuzz build --features libfuzzer-sys
```

I tweaked the $DATE variable to try and match the pinned zerocopy nightly toolchain, but the error remained unchanged.

Oh hahaha I totally missed the crucial piece of information in your original message:

```text
error[E0080]: evaluation of `<[()] as FromBytes::ref_from_bytes::StaticAssert>::ASSERT` failed
```

Unless I'm missing something, the call in question is to `<[()] as FromBytes>::ref_from_bytes`.

> I tried upgrading to v0.8.3 without adding `KnownLayout` to any types [(ref)](https://github.com/clundin25/caliptra-dpe/commit/7716fa79103d5f1fcb487a79af542fd174ef734a) and the issue persisted, which makes me believe my code changes are not the root cause, since I haven't accidentally decorated a DST with a zero-sized tail.

You wouldn't need to explicitly add `KnownLayout` - it's already implemented for some built-in types (such as `[()]`). I believe that this error message is not about a type which *contains* a `[()]`, but rather about `[()]` itself as the top-level type being passed to `FromBytes::ref_from_bytes`:

```text
error[E0080]: evaluation of `<[()] as FromBytes::ref_from_bytes::StaticAssert>::ASSERT` failed
```

Unfortunately it's closed-source, but I've found what seems to be a minimal reproduceable example.

`Cargo.lock`:
```
[package]
name = "zc-dead-code"
version = "0.1.0"
edition = "2021"

[dependencies]
zerocopy = { version="0.8.5", features=["derive"] }
```

`lib.rs`:
```
use zerocopy::{FromBytes, IntoBytes};

#[derive(IntoBytes, FromBytes)]
pub struct Foo;
```

On Windows (I've tested using both cargo/rustc versions 1.78 and 1.81), `cargo rustc` successfully builds, whereas `cargo rustc -- -Clink-dead-code` returns two errors:
```
error[E0080]: evaluation of `<[Foo] as zerocopy::FromBytes::mut_from_bytes::StaticAssert>::ASSERT` failed
    --> [path_redacted]\zerocopy-0.8.5\src\lib.rs:3744:9
     |
3744 |         static_assert_dst_is_not_zst!(Self);
     |         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ the evaluated program panicked at 'assertion failed: {
    let dst_is_zst =
        match Self::LAYOUT.size_info {
            crate::SizeInfo::Sized { .. } => false,
            crate::SizeInfo::SliceDst(TrailingSliceLayout { elem_size, .. })
                => {
                elem_size == 0
            }
        };
    !dst_is_zst
}', [path_redacted]\zerocopy-0.8.5\src\lib.rs:3744:9
     |
     = note: this error originates in the macro `assert` which comes from the expansion of the macro `static_assert_dst_is_not_zst` (in Nightly builds, run with -Z macro-backtrace for more info)

note: erroneous constant encountered
    --> [path_redacted]\zerocopy-0.8.5\src\lib.rs:3744:9
     |
3744 |         static_assert_dst_is_not_zst!(Self);
     |         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
     |
     = note: this note originates in the macro `static_assert` which comes from the expansion of the macro `static_assert_dst_is_not_zst` (in Nightly builds, run with -Z macro-backtrace for more info)   

note: the above error was encountered while instantiating `fn <[Foo] as zerocopy::FromBytes>::mut_from_bytes`
    --> [path_redacted]\zerocopy-0.8.5\src\lib.rs:4588:9
     |
4588 |         <[Self]>::mut_from_bytes(source).ok()
     |         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

For more information about this error, try `rustc --explain E0080`.
```

--- a/Cargo.toml
+++ b/Cargo.toml
@@ -15,7 +15,7 @@
 [package]
 edition = "2021"
 name = "zerocopy"
-version = "0.8.5"
+version = "0.8.6"
 authors = ["Joshua Liebow-Feeser <joshlf@google.com>"]
 description = "Zerocopy makes zero-cost memory manipulation effortless. We write \"unsafe\" so you don't have to."
 categories = ["embedded", "encoding", "no-std::no-alloc", "parsing", "rust-patterns"]
@@ -77,13 +77,13 @@ std = ["alloc"]
 __internal_use_only_features_that_work_on_stable = ["alloc", "derive", "simd", "std"]
 
 [dependencies]
-zerocopy-derive = { version = "=0.8.5", path = "zerocopy-derive", optional = true }
+zerocopy-derive = { version = "=0.8.6", path = "zerocopy-derive", optional = true }
 
 # The "associated proc macro pattern" ensures that the versions of zerocopy and
 # zerocopy-derive remain equal, even if the 'derive' feature isn't used.
 # See: https://github.com/matklad/macro-dep-test
 [target.'cfg(any())'.dependencies]
-zerocopy-derive = { version = "=0.8.5", path = "zerocopy-derive" }
+zerocopy-derive = { version = "=0.8.6", path = "zerocopy-derive" }
 
 [dev-dependencies]
 itertools = "0.11"
@@ -97,6 +97,6 @@ testutil = { path = "testutil" }
 # CI test failures.
 trybuild = { version = "=1.0.89", features = ["diff"] }
 # In tests, unlike in production, zerocopy-derive is not optional
-zerocopy-derive = { version = "=0.8.5", path = "zerocopy-derive" }
+zerocopy-derive = { version = "=0.8.6", path = "zerocopy-derive" }
 # TODO(#381) Remove this dependency once we have our own layout gadgets.
 elain = "0.3.0"

--- a/src/lib.rs
+++ b/src/lib.rs
@@ -4544,17 +4544,6 @@ pub unsafe trait FromBytes: FromZeros {
         Self::mut_from_bytes(source).ok()
     }
 
-    #[deprecated(since = "0.8.0", note = "`FromBytes::ref_from_bytes` now supports slices")]
-    #[doc(hidden)]
-    #[must_use = "has no side effects"]
-    #[inline(always)]
-    fn slice_from(source: &[u8]) -> Option<&[Self]>
-    where
-        Self: Sized + Immutable,
-    {
-        <[Self]>::ref_from_bytes(source).ok()
-    }
-
     #[deprecated(since = "0.8.0", note = "renamed to `FromBytes::ref_from_prefix_with_elems`")]
     #[doc(hidden)]
     #[must_use = "has no side effects"]
@@ -4575,17 +4564,6 @@ pub unsafe trait FromBytes: FromZeros {
         Self: Sized + Immutable,
     {
         <[Self]>::ref_from_suffix_with_elems(source, count).ok()
-    }
-
-    #[deprecated(since = "0.8.0", note = "`FromBytes::mut_from_bytes` now supports slices")]
-    #[must_use = "has no side effects"]
-    #[doc(hidden)]
-    #[inline(always)]
-    fn mut_slice_from(source: &mut [u8]) -> Option<&mut [Self]>
-    where
-        Self: Sized + IntoBytes,
-    {
-        <[Self]>::mut_from_bytes(source).ok()
     }
 
     #[deprecated(since = "0.8.0", note = "renamed to `FromBytes::mut_from_prefix_with_elems`")]

--- a/zerocopy-derive/Cargo.toml
+++ b/zerocopy-derive/Cargo.toml
@@ -9,7 +9,7 @@
 [package]
 edition = "2021"
 name = "zerocopy-derive"
-version = "0.8.5"
+version = "0.8.6"
 authors = ["Joshua Liebow-Feeser <joshlf@google.com>"]
 description = "Custom derive for traits from the zerocopy crate"
 license = "BSD-2-Clause OR Apache-2.0 OR MIT"

